src/Psl/DateTime/DateTimeConvenienceMethodsTrait.php

```diff
@@ -0,0 +1,595 @@
+     *
+     * @return array{int<1, 12>, Meridiem} The hours and meridiem indicator.
+     *
+     * @psalm-mutation-free
+     */
+    public function getTwelveHours(): array
+    {
+        return [
+            ($this->getHours() % 12 ?: 12),
+            ($this->getHours() < 12 ? Meridiem::AnteMeridiem : Meridiem::PostMeridiem),
+        ];
+    }
+
+    /**
+     * Retrieves the ISO-8601 year and week number corresponding to the date.
+     *
+     * This method returns an array consisting of two integers: the first represents the year, and the second
+     * represents the week number according to ISO-8601 standards, which ranges from 1 to 53. The week numbering
+     * follows the ISO-8601 specification, where a week starts on a Monday and the first week of the year is the
+     * one that contains at least four days of the new year.
+     *
+     * Due to the ISO-8601 week numbering rules, the returned year might not always match the Gregorian year
+     * obtained from `$this->getYear()`. Specifically:
+     *
+     *  - The first few days of January might belong to the last week of the previous year if they fall before
+     *      the first Thursday of January.
+     *
+     *  - Conversely, the last days of December might be part of the first week of the following year if they
+     *      extend beyond the last Thursday of December.
+     *
+     * Examples:
+     *  - For the date 2020-01-01, it returns [2020, 1], indicating the first week of 2020.
+     *  - For the date 2021-01-01, it returns [2020, 53], showing that this day is part of the last week of 2020
+     *      according to ISO-8601.
+     *
+     * @return array{int, int<1, 53>}
+     *
+     * @psalm-mutation-free
+     */
+    public function getISOWeekNumber(): array
+    {
+        /** @var int<1, 53> $week */
+        $week = (int)$this->format(pattern: 'w');
+        $year = (int)$this->format(pattern: 'Y');
+
+        return [$year, $week];
+    }
+
+    /**
+     * Gets the weekday of the date.
+     *
+     * @return Weekday The weekday.
+     *
+     * @psalm-mutation-free
+     */
+    public function getWeekday(): Weekday
+    {
+        return Weekday::from((int) $this->format(pattern: 'e'));
+    }
+
+    /**
+     * Checks if the year is a leap year.
+     *
+     * @psalm-mutation-free
+     */
+    public function isLeapYear(): bool
+    {
+        return namespace\is_leap_year($this->getYear());
+    }
+
+    /**
+     * Adds the specified years to this date-time object, returning a new instance with the added years.
+     *
+     * @throws Exception\UnderflowException If adding the years results in an arithmetic underflow.
+     * @throws Exception\OverflowException If adding the years results in an arithmetic overflow.
+     *
+     * @psalm-mutation-free
+     */
+    public function plusYears(int $years): static
+    {
+        return $this->plusMonths($years * MONTHS_PER_YEAR);
```

Comment:
fixed.

Reply:
> Duration::years(18) but it seems like Duration stops at weeks.
Does it make sense to add more methods to duration (months, years)

Yes, that is not possible due to leap years, and months have different length, e.g we don't know if Duration::month(1) should return 28 days, 29, 30, or 31.

Reply:
same for years.

Reply:
I tried the fix with leap days:

```
$at = \Psl\DateTime\DateTime::parse('2024-02-29 09:00:00', FormatPattern::SqlDateTime)->minusYears(18);
echo $at->toRfc3339();
```

This results in:

> 2006-03-01T09:00:00+00:00

Which is also what PHP is giving you with regular DateTime.
However,
I'dd rather expect it to be `2006-02-28T09:00:00+00:00` - since march did not start yet.

This is how brick/date-time does it:

https://github.com/brick/date-time/blob/4682c4e1c3cd0b39cce6ac2d6c5eee60de27bde2/src/LocalDate.php#L817-L819

Example:

```php
$today = new DateTimeImmutable('2024-02-29');
$nextYear = $today->modify('+1 year');
echo $nextYear->format('Y-m-d') . PHP_EOL; // 2025-03-01

$today = \Brick\DateTime\LocalDate::fromNativeDateTime($today);
$nextYear = $today->plusYears(1);

echo $nextYear->__toString() . PHP_EOL;  // 2025-02-28
```

(It's based on java's : JSR 310 (Date and Time API))

Reply:
this is really depends on the library/api, there is no standard telling us to go a day forward, or a day back.

I suppose going back a day makes sense as it leaves you within the same month 🤔

Reply:
case for 28-02: we are still in february, the last day.
case for 01-03: 28 days have already passed in february, we are in the day after the 28th, since its not a leap year, its 01-03.

Reply:
It's only an issue if you add a year or more on a leap day though.

So a better question to ask is: what should one expect if you add a year to a leap day, just like how should adding a month on the 31th of January behave

Reply:
... which brings me to this:

```php
$d = new \DateTimeImmutable('2024-01-31 09:00:00');
echo $d->add(\DateInterval::createFromDateString('1 month'))->format('Y-m-d');
```

> 2024-03-02


Java / brick:
```java
LocalDate leapDay = LocalDate.of(2024, Month.JANUARY, 31);
Period oneMonth = Period.ofMonths(1);
LocalDate added = leapDay.plus(oneMonth);
```

> 2024-02-29


To me, it makes sense to take the java path here as well.